mm2src/mm2_main/src/lp_healthcheck.rs

```diff
@@ -0,0 +1,513 @@
+}
+
+fn deserialize_bytes<'de, D>(deserializer: D) -> Result<Vec<u8>, D::Error>
+where
+    D: serde::Deserializer<'de>,
+{
+    struct ByteVisitor;
+
+    impl<'de> serde::de::Visitor<'de> for ByteVisitor {
+        type Value = Vec<u8>;
+
+        fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+            formatter.write_str("a non-empty byte array up to 512 bytes")
+        }
+
+        fn visit_seq<A>(self, mut seq: A) -> Result<Vec<u8>, A::Error>
+        where
+            A: serde::de::SeqAccess<'de>,
+        {
+            let mut buffer = vec![];
+            while let Some(byte) = seq.next_element()? {
+                if buffer.len() >= 512 {
+                    return Err(serde::de::Error::invalid_length(
+                        buffer.len(),
+                        &"longest possible length allowed for this field is 512 bytes (with RSA algorithm).",
+                    ));
+                }
+
+                buffer.push(byte);
+            }
+
+            if buffer.is_empty() {
+                return Err(serde::de::Error::custom("Can't be empty."));
+            }
+
+            Ok(buffer)
+        }
+    }
+
+    deserializer.deserialize_seq(ByteVisitor)
+}
+
+#[derive(Debug, Display, Serialize, SerializeErrorType)]
+#[serde(tag = "error_type", content = "error_data")]
+pub enum HealthcheckRpcError {
+    MessageGenerationFailed { reason: String },
+    MessageEncodingFailed { reason: String },
+    Internal { reason: String },
+}
+
+impl HttpStatusCode for HealthcheckRpcError {
+    fn status_code(&self) -> common::StatusCode {
+        match self {
+            HealthcheckRpcError::MessageGenerationFailed { .. }
+            | HealthcheckRpcError::Internal { .. }
+            | HealthcheckRpcError::MessageEncodingFailed { .. } => StatusCode::INTERNAL_SERVER_ERROR,
+        }
+    }
+}
+
+pub async fn peer_connection_healthcheck_rpc(
+    ctx: MmArc,
+    req: RequestPayload,
+) -> Result<bool, MmError<HealthcheckRpcError>> {
+    // When things go awry, we want records to clear themselves to keep the memory clean of unused data.
+    // This is unrelated to the timeout logic.
+    let address_record_exp = Duration::from_secs(healthcheck_message_exp_secs());
+
+    let target_peer_address = req.peer_address;
+
+    let p2p_ctx = P2PContext::fetch_from_mm_arc(&ctx);
+    if target_peer_address == p2p_ctx.peer_id().into() {
+        // That's us, so return true.
+        return Ok(true);
+    }
+
+    let message = HealthcheckMessage::generate_message(&ctx, false)
+        .map_err(|reason| HealthcheckRpcError::MessageGenerationFailed { reason })?;
+
+    let encoded_message = message
+        .encode()
+        .map_err(|e| HealthcheckRpcError::MessageEncodingFailed { reason: e.to_string() })?;
+
+    let (tx, rx): (Sender<()>, Receiver<()>) = oneshot::channel();
+
+    {
+        let mut book = ctx.healthcheck_response_handler.lock().await;
+        book.clear_expired_entries();
```

Comment:
This `clear_expired_entries` will be removed when merging with this https://github.com/KomodoPlatform/komodo-defi-framework/pull/2232 depending on which gets merged first. Just a note :)